### PR TITLE
extend VectorTools to complex-valued PETSc vector

### PR DESCRIPTION
**Pull request by davydden**
- [x] fix mean value (ideas?)
- [x] avoid `std::norm()`

in reference to https://github.com/dealii/dealii/issues/2033

### CODE REVIEW

File: include/deal.II/numerics/vector_tools.templates.h [hunk 1]
```diff
@@ -6356,7 +6359,7 @@ namespace VectorTools
 
       // append result of this cell to the end of the vector
       AssertIsFinite(diff);
-      return diff;
+      return diff; // TODO: what about diff_mean? Not necessarily double
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: What remains to be done here?

**Comment by bangerth** [comment]
Some of these changes break backward compatibility because one could previously pass in a `Vector<double>` as input argument and `Vector<double>` as output argument. Now they need to have the same underlying type. Which is how it should be, but it is incompatible.

Can you add a blurb in the "incompatibilities" section of changes.h pointing this out?

File: include/deal.II/numerics/vector_tools.templates.h (hunk 1 above)
Comment (davydden) [inline include/deal.II/numerics/vector_tools.templates.h]: i actually don't know ;-) the point is that the norm should be real-valued. In that sense, the mean-value is not a norm, actually. I guess, i will add an assert and return real-part of the mean value for complex-valued vectors. Does this sound ok to you?

Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: Yes, that works. I think I simply didn't understand from that one sentence what the TODO was about. In either case, document your choice.

**Comment by davydden** [comment]
i will add a note to changes.h after another PR is merged with the similar incompatibilities.

File: include/deal.II/numerics/vector_tools.templates.h [hunk 2]
```diff
@@ -6102,6 +6104,25 @@ namespace VectorTools
...
+    {
+      template<typename number>
+      double mean_to_double(const number &mean_value)
+      {
+        return mean_value;
+      }
+
+      template<typename number>
+      double mean_to_double(const std::complex<number> &mean_value)
+      {
```
Comment (bangerth) [inline include/deal.II/numerics/vector_tools.templates.h]: Ah, yes, this is good.

**Comment by bangerth** [comment]
OK, please make the change and merge yourself.

**Pull request merged**